packages/contracts-bedrock/src/interop/L2CrossDomainMessenger.sol

```diff
@@ -0,0 +1,186 @@
+    function sendMessage(
+        bytes32 _destination,
+        address _target,
+        bytes calldata _message,
+        uint32 _minGasLimit
+    )
+        external
+        payable
+    {
+        // L2->L1 Support: Required as we are leaving the CrossDomainMessenger untouched
+        bytes32 ETH_MAINNET_ID = bytes32(hex"01");
+        if (_destination == ETH_MAINNET_ID) {
+            LegacyL2CrossDomainMessenger(Predeploys.L2_CROSS_DOMAIN_MESSENGER).sendMessage(
+                _target, _message, _minGasLimit
+            );
+            return;
+        }
+
+        require(_destination != CHAIN_ID, "L2CrossDomainMessenger: message cant be sent to self");
+
+        uint256 nonce = Encoding.encodeVersionedNonce(msgNonce, MESSAGE_VERSION);
+        bytes memory data = abi.encodeWithSelector(
+            this.relayMessage.selector, nonce, CHAIN_ID, msg.sender, _target, msg.value, _minGasLimit, _message
+        );
+
+        // (1) Send to the CrossL2Inbox.
+        // With an updated CrossDomainMessenger message format, this contract will
+        // replace existing L2CrossDomainmMessenger predeploy and switch between
+        // L2ToL1MessagePasser/CrossL2Inbox based on the provided `_destination`.
+        //
+        // i.e: CrossL2Inbox.sendMessage(source, destination, _minGasLimit, msg.value, data)
+
+        // (2) Emit Events
+
+        unchecked {
+            ++msgNonce;
+        }
+    }
+
+    /// @notice Relays a message that was sent from a different chain. Since L1-L2
+    ///         messages utilize the V1 message format, this entrypoint only
+    ///         supports relaying L2-L2 messages.
+    function relayMessage(
+        uint256 _nonce,
+        bytes32 _source,
+        address _sender,
+        address _target,
+        uint256 _value,
+        uint256 _minGasLimit,
+        bytes calldata _message
+    )
+        external
+        payable
+    {
+        (, uint16 version) = Encoding.decodeVersionedNonce(_nonce);
+        require(version == MESSAGE_VERSION, "L2CrossDomainMessenger: incorrect message version");
+
+        bytes32 messageHash =
+            hashCrossDomainMessageV2(_nonce, _source, CHAIN_ID, _sender, _target, _value, _minGasLimit, _message);
+        require(successfulMessages[messageHash] == false, "L2CrossDomainMessenger: message already processed");
+
+        // (1) Allow for replay
+        if (_sender == address(this)) {
+            assert(msg.value == _value);
+            assert(failedMessages[messageHash] == false);
+        } else {
+            require(msg.value == 0, "cannot replay with more funds");
+            require(failedMessages[messageHash], "L2CrossDomainMessengeR: message cannot be replayed");
+        }
+
+        // **CrossDomainMessenger Checks**. min gas, unsafe target, etc.
+
+        // (2) Relay Message
+        xDomainMsgSender = _sender;
+        bool success = SafeCall.call(_target, gasleft(), _value, _message);
```

Comment:
nit: try to use kwargs syntax when possible to ensure arguments are in the correct order